Manager/src/main/java/io/raspberrywallet/manager/bitcoin/Bitcoin.java

```diff
@@ -51,58 +80,119 @@ private NetworkParameters parseNetworkFrom(Configuration.BitcoinConfig bitcoinCo
+                wallet = Wallet.loadFromFile(walletFile);
+
+                if (!wallet.isEncrypted()) {
+                    saveEncryptedWallet(key);
+                    throw new SecurityException("Decrypted wallet on disk detected");
+                } else
+                    decryptWallet(key);
+
+                synchronizeWalletBlocking(wallet);
+                saveEncryptedWallet(key);
+                Logger.info("Wallet balance" + wallet.getBalance().toFriendlyString());
+            } catch (IOException | BlockStoreException | UnreadableWalletException | WalletNotInitialized walletNotInitialized) {
+                walletNotInitialized.printStackTrace();
             }
         };
-        // Now configure and start the appkit. This will take a second or two - we could show a temporary splash screen
-        // or progress widget to keep the user engaged whilst we initialise, but we don't.
-        kit.setAutoSave(false)
-                .setBlockingStartup(false)
-                .setUserAgent("RaspberryWallet", "1.0");
-
-        if (seed != null)
-            kit.restoreWalletFromSeed(seed);
+        if (peerGroup != null && peerGroup.isRunning()) {
+            // Shut down synchronization before setting up new wallet
+            ListenableFuture future = peerGroup.stopAsync();
+            future.addListener(setupWalletFromBackup, Executors.newSingleThreadExecutor());
+        } else
+            setupWalletFromBackup.run();
     }
 
-    public WalletAppKit getKit() throws WalletNotInitialized {
-        if (kit == null) throw new WalletNotInitialized();
-        return kit;
+
+    /**
+     * Download blockchain from PeerGroup discovered by DnsDiscovery on the NetworkParams
+     *
+     * @param wallet wallet to index transactions for
+     */
+    private void synchronizeWalletBlocking(Wallet wallet) throws BlockStoreException, IOException {
+        long start = System.currentTimeMillis();
+
+        InputStream checkpoints = CheckpointManager.openStream(params);
+        CheckpointManager.checkpoint(params, checkpoints, blockStore, wallet.getEarliestKeyCreationTime());
+
+        BlockChain chain = new BlockChain(params, wallet, blockStore);
+
+        peerGroup = new PeerGroup(params, chain);
+        peerGroup.setUserAgent("RaspberryWallet", "1.0");
+        peerGroup.addPeerDiscovery(new DnsDiscovery(params));
+        peerGroup.addAddress(new PeerAddress(params, InetAddress.getLocalHost()));
+        peerGroup.start();
+        peerGroup.downloadBlockChain();
+
+        long total = System.currentTimeMillis() - start;
+        Logger.info(String.format("Synchronization took %.2f secs", (double) total / 1000.0));
     }
 
+    @NotNull
     public Wallet getWallet() throws WalletNotInitialized {
-        if (kit == null) throw new WalletNotInitialized();
-        return kit.wallet();
+        Wallet _wallet = wallet;
+        if (_wallet == null) throw new WalletNotInitialized();
+        return _wallet;
```

Comment:
could be simplyfied to:
```
return wallet == null ? throw new WalletNotInitialized() : wallet;
```

Reply:
Are you sure that trinary operator is atomic ?

Reply:
I'm almost sure it's not.

Reply:
but does it matter? there are 3 lines that doesn't execute as atomically

Reply:
I approve changes, if it doesn't matter, just merge it into master

Reply:
By atomicaly I mean thread race safety. If there is space for another thread between check `wallet == null` and `return wallet;` to change the value to null.